File: catkin_ws/src/motor_driver_library/src/MotorDriver.py
import Jetson.GPIO as GPIO
import 

class Motor:
    def __init__(self, directionPin, pwmPin, minSpeed, maxSpeed):
        self.m_directionPin = directionPin
        self.m_pwmPin = pwmPin
        self.m_minSpeed = minSpeed
        self.m_maxSpeed = maxSpeed
        self.m_speed = 0.0
        self.m_direction = 0

        # Pin configurations for JetsonGPIO
        GPIO.setmode(GPIO.BOARD)  # GPIO.BCM or GPIO.BOARD
        GPIO.setup(32, GPIO.OUT, initial=GPIO.HIGH)  # Left motor
        GPIO.setup(33, GPIO.OUT, initial=GPIO.HIGH)  # Right motor
        GPIO.setup(31, GPIO.OUT, initial=GPIO.HIGH)  # Left direction
        GPIO.setup(35, GPIO.OUT, initial=GPIO.HIGH)  # Right direction

    # --- Pwm pin ---
    def getPwmPin(self):
        return self.m_pwmPin
    # --- End of Pwm pin ---

    # --- Wheel direction ---
    def setDirection(self, speed):
        if speed >= 0:
            self.m_direction = True
        elif speed < 0:
            self.m_direction = False

    def getDirection(self):
        return self.m_direction

    def getDirectionPin(self):
        return self.m_directionPin
    # --- End of wheel direction ---

    # --- Control of motor actions ---
    def setSpeed(self, speed):
        if speed <= self.m_minSpeed:
            self.m_speed = self.m_minSpeed
        elif speed >= self.m_maxSpeed:
            self.m_speed = self.m_maxSpeed
        else:
            self.m_speed = speed

    def getSpeed(self):
        return self.m_speed

    def setMinSpeed(self, minSpeed):
        self.m_minSpeed = minSpeed

    def setMaxSpeed(self, maxSpeed):
        self.m_maxSpeed = maxSpeed

    def stop(self):
        self.m_speed = 0.0
    # --- End of control of motor actions ---

File: library/MotorDriver/src/MotorDriver.cpp
#include "MotorDriver.h"

MD::Motor::Motor(int directionPin, int pwmPin, double minSpeed, double maxSpeed) {
    m_directionPin = directionPin;
    m_pwmPin = pwmPin;
    m_minSpeed = minSpeed;
    m_maxSpeed = maxSpeed;
    m_speed = 0.0;
    m_direction = 0;

	// Pin configurations for JetsonGPIO
	GPIO::setmode(GPIO::BOARD); // GPIO::BCM or GPIO::BOARD
	GPIO::setup(32, GPIO::OUT, GPIO::HIGH);
	//GPIO::setup(12, GPIO::OUT);
   
}

// --- Pwm pin ---.
int MD::Motor::getPwmPin(){
    return m_pwmPin;
}
// --- End of Pwm pin ---

// --- Wheel direction ---
void MD::Motor::setDirection(int direction) {
    m_direction = direction;
    if (direction == 0) {
        GPIO::output(31, GPIO::HIGH);
    } else {
        GPIO::output(31, GPIO::LOW);
    }
}

int MD::Motor::getDirection() {return m_direction;}

int MD::Motor::getDirectionPin(){
    return m_directionPin;
}
// --- End of wheel direction ---

// --- Control of motor actions ---
void MD::Motor::setSpeed(double speed) {
    if (speed <= m_minSpeed) {
        m_speed = m_minSpeed;
    } else if (speed >= m_maxSpeed) {
        m_speed = m_maxSpeed;
    } else {
    	m_speed = speed;
    }
}

double MD::Motor::getSpeed() {
    return m_speed;
}

void MD::Motor::setMinSpeed(double minSpeed) {
    m_minSpeed = minSpeed;
}

// int Motor::getMinSpeed() {return 0;}

void MD::Motor::setMaxSpeed(double maxSpeed) {
    m_maxSpeed = maxSpeed;
}

// int Motor::getMaxSpeed() {return 0;}

/*
void MD::Motor::startMotor(int speed) {
    if (speed < m_minSpeed) {
        speed = m_minSpeed;
    } else if (speed > m_maxSpeed) {
        speed = m_maxSpeed;
    }

    if (m_direction) {
        GPIO::output(m_directionPin, GPIO::HIGH);
    } else {
        GPIO::output(m_directionPin, GPIO::LOW);
    }
}
*/

void MD::Motor::stop() {
    m_speed = 0.0;
}
// --- End of control of motor actions ---

